cpp/docs/DOCUMENTATION.md

```diff
@@ -0,0 +1,453 @@
+Also include append `[in]`, `[out]` or `[in,out]` to the `@param` if it is not clear from the declaration and the parameter name itself.
+
+```c++
+ *
+ * @param[in]     first  This parameter is an input parameter to the function
+ * @param[in,out] second This parameter is used both as an input and output
+ * @param[out]    third  This parameter is an output of the function
+ *
+```
+
+Also recommend vertically aligning the 3 columns of text if possible to make it easier to read in the source code itself.
+
+#### @return
+
+Add a single [@return](http://www.doxygen.nl/manual/commands.html#cmdreturn) comment line at the end of the comment block if the function returns an object or value.
+Include a brief description of what is returned.
+
+```c++
+/**
+ * ...
+ *
+ * @return A new column of type INT32 and no nulls
+ */
+```
+
+Do not include the type of the object returned with the `@return` comment.
+
+### Inline Examples
+
+It is usually helpful to include a source code example inside your comment block when documenting a function or other declaration.
+Use the [@code](http://www.doxygen.nl/manual/commands.html#cmdcode) and [@endcode](http://www.doxygen.nl/manual/commands.html#cmdendcode) pair to include inline examples.
+
+Doxygen supports syntax highlighting for C++ and several other programming languages (e.g. Python, Java).
+By default, the `@code` tag will use syntax highlighting based on the source code where it was found.
+
+```c++
+ *
+ * @code
+ * auto result = cudf::make_column( );
+ * @endcode
+ *
+```
+
+You can specify a different language by indicating the file extension in the tag:
+
+```c++
+ *
+ * @code{.py}
+ * import cudf
+ * s = cudf.Series([1,2,3])
+ * @endcode
+ *
+```
+
+If you wish to use pseudo-code in your example, use the following:
+
+```c++
+ *
+ * Sometimes pseudo-code is clearer.
+ * @code{.pseudo}
+ * s = int column of [ 1, 2, null, 4 ]
+ * r = fill( s, [1, 2], 0 )
+ * r is now [ 1, 0, 0, 4 ]
+ * @endcode
+ *
+```
+
+When writing example snippets, using fully qualified class names allows doxygen to add reference links to the example.
+
+```c++
+ *
+ * @code
+ * auto result1 = make_column( ); // reference link will not be created
+ * auto result2 = cudf::make_column( ); // reference link will be created
+ * @endcode
+ *
+```
+
+Although using 3 backtick marks `` ``` `` for example blocks will work too, they do not standout as well in vs-code and other source editors.
+
+Do not use the `@example` tag in the comments for a declaration.
+Doxygen will interpret the entire source file as example source code when it encounters this tag.
+The source file is then published under a separate the _Examples_ page in the output.
+
+## Namespaces
+
+Doxygen output includes a _Namespaces_ page that shows all the namespaces declared with comment blocks in the processed files.
+Here is an example of doxygen description comment for a namespace declaration.
+
+```c++
+//! cuDF interfaces
```

Comment:
Shouldn't any namespace we document include more detail and therefore deserve a `/**` doxygen block?